### ISSUE TITLE
Initializing Wren attributes from C

### ISSUE THREAD

**Issue by Skyway666** [wren-lang/wren#639]
I am building the base for a game Engine as a subject in my degree. My subsystem of choice was scripting, and the language Wren. 

I red the documentation of "Embedding", and I managed to understand how to call Wren from C and viceversa. Even then, if I didn't overlook it, it doesn't seem to tackle the issue of initializing variables from a Wren class in C.

This is crucial, since, if I make a "MoveWithArrows" script, I might want to change the variable "speed" of the class and have different GameObjects being controlled with the arrows at different speeds, similar to how unity works... Can anybody help me?

**Comment by Skyway666** [comment]
But even then, if a user creates a getter and a setter, I need to be able to get the name of it from the script, in order to create the handler. Isn't there a way to get all the signatures from a class in some sort of list?

The only idea that i can come up with is to have a file attached to the script, a json for instances, with the name of the variables that need to be set in the editor, and then generate the appropiate handlers reading it, and invoque the getters before start executing the "Update" function in the class.

**Comment by underscorediscovery** [comment]
There is no API provided by the public facing API right now to fetch a list of methods.

You can reach past the API, if you want to, and get a list directly.  You'll need to import `wren_value.h` at least to interact with the vm types directly. 

Here's an example of a foreign function that returns a list of fields on a given class. It's a foreign for example sake, but you could push the strings into a vector/map or other container as needed.

```
void get_class_fields(WrenVM* vm) {

    WrenHandle* handle = wrenGetSlotHandle(vm, 1);
  
    wrenEnsureSlots(vm, 2);
    wrenSetSlotNewList(vm, 0);
    
      //important part is here
    if(IS_CLASS(handle->value)) {
      ObjClass* cls = AS_CLASS(handle->value);
      printf("%.*s\n", cls->name->length, cls->name->value);
      for(int i = 0; i < cls->methods.count; ++i) {
        Method& method = cls->methods.data[i];
        if(method.type != METHOD_PRIMITIVE && method.type != METHOD_NONE) {
          wrenSetSlotString(vm, 1, method.fn.obj->fn->debug->name);
          wrenInsertInList(vm, 0, -1, 1);
        }
      }
    }

    wrenReleaseHandle(vm, handle);

} //get_class_fields
```

**Comment by underscorediscovery** [comment]
No problem. 
If using wren, you can use `signature.endsWith("=(_)")` to find setters, and something like `!signature.contains("(") && !signature.contains("[")` to find getters. Or maybe `inValidIdentifierRange(signature[-1])`. Valid identifier range is an example, but would use the wren syntax rules to find ones that end without any decoration.

**Comment by Skyway666** [comment]
It worked wonders. Now I can go on with my project, I'm so glad!

I think this functionality should make it to the official API, it's very useful.

The code you gave me has 3 little issues I found, I managed to fix 2 and the other doesn't affect me though, but I hope this is helpful to whoever has the same problem than me

- For some reason it won't get the static functions (I couldn't fix it)

- If the class has any foreigner functions, it will crash trying to get the name, unless you check that the type is not foreigner before
- The code to get the name of the signature from the method won't compile :(

This is my final implementation:

```
std::vector<std::string> ModuleScripting::GetMethodsFromClassHandler(WrenHandle * wrenClass) {

	if (!IS_CLASS(wrenClass->value)) {
		app_log->AddLog("Trying to get methods from a non class handler");
		return std::vector<std::string>();
	}
	std::vector<std::string> ret;
	ObjClass* cls = AS_CLASS(wrenClass->value);
	app_log->AddLog("%.*s\n", cls->name->length, cls->name->value);
	for (int i = 0; i < cls->methods.count; ++i) {
		Method& method = cls->methods.data[i];
		if (method.type != METHOD_PRIMITIVE && method.type != METHOD_FOREIGN && method.type != METHOD_NONE ) {
			ret.push_back(method.as.closure->fn->debug->name);
		}
	}

	return ret;
}
```

**Comment by Skyway666** [comment]
So... It seems like I got excited too early. I was able to get the methods from a class, including the setters, but when I try to call it, wren sends an error:

 "console_test metaclass does not implement 'speed=(_)' error in -1 line"

this is my wren code 

```
class EngineComunicator{
	construct new(){}
	foreign static consoleOutput(message)
}

class console_test{
	speed{ _speed }
	speed=(v) {_speed = v }
	 construct new() {
		_speed = 0
	}
	static Update(){
		EngineComunicator.consoleOutput("Hello world from a wren function")
	}
	addition(a, b){
		return a + b
	}
}

```

The class handler is seemingly being generated correctly, because I'm able to call the "Update" function. 

The signature handler as well, since it is giveing me the right signature when the error is sent, besides is one line of code, there is not much I can mess up 

```
void ComponentScript::GenerateSetterHandlers() {

	for (int i = 0; i < editor_variables.size(); i++) {
		std::string setter_signature = editor_variables[i].getName();
		setter_signature += "=(_)";
		setter_handlers.push_back(wrenMakeCallHandle(App->scripting->vm, setter_signature.c_str()));
	}
} 
```

Can you please help me?

**Comment by underscorediscovery** [comment]
`console_test metaclass` is telling you that you're calling a method _on the class itself_. In simpler terms, `static speed=(_)`. You need to put an _instance_ of the class into slot 0, then call the instance method. If you put the _class itself_ it will be a static.

**Comment by Skyway666** [comment]
Well... I was planning on calling static methods for my scripting system, I hoped to change the variables for the class before calling "Update()" or "Start()", since in the documentation says that "most of the time you'll be calling static wren functions...". Maybe it is a terrible approach on my side.

Can I create a static setter-getter? and if I can't, how do I get a handler to a class instance? In the documentation, the only way it explains to get a handler is through this code:

```
wrenEnsureSlots(vm, 1); 
wrenGetVariable(vm, "main", "GameEngine", 0); 
WrenHandle* gameEngineClass = wrenGetSlotHandle(vm, 0);
```

I'm gessing this gives you "the class itself"

Thank you very much for your help anyways, I really wanna make this work!

**Comment by underscorediscovery** [comment]
It's usually pretty quick to just try things yourself, for the basic questions :) I don't mind helping, but you can save yourself time by reading the docs and just experimenting.

```js
class Test {
  static hi=(v) { __hi = v }
  static hi { "hi `%(__hi)`" }
}
```
with
```js
Test.hi = "there"
System.print(Test.hi)
```

**Comment by underscorediscovery** [comment]
On the second part, the example code is fetching a variable. That variable can be a class `class GameEngine` or it can be a variable, `var GameEngine = MyEngine.new()`. The caller is just asking for a variable by name. It's up to you to configure the details of whether it's static or not.

You can change the thing that Update gets called on, this is done with the "receiver", which is in slot 0. If the receiver is a class it goes to static methods, if it's an instance it goes to that instance. If you have a list of WrenHandle*, you should be able to just call this in a loop and each time put the unique instance into slot 0 before calling.

**Comment by Skyway666** [comment]
The static setter/getter would work, I tried to put the keyword in front of the method, but I didn't know you needed to put two underscores, silly me :-1: 

Anyways, now I run into another problem which I thought was non important... The function that you provided me to get methods from a class doesn't provide the static ones. That wasn't an issue before because I thought I didn't need to make setters/getters static, but now is a problem...

**Comment by underscorediscovery** [comment]
If you're trying to do similar to how something like Unity calls `Update` for a script, it's probably best workflow wise, that the functions are instance methods and not static. The user would likely want to have local variables to the class, and forcing them to go via statics + module level variables is less fun.

Let's say for example you made this your script definition:
```js
class Script {
   construct new() {} //required
   Update() {} //can be optional if you're testing existence
}
```

What you can try is:
- get the class
  - i.e `wrenGetVariable(vm, "main", "Script", 0);` 
    - note the module 'main' here, it would likely be specific to the script
- make a new instance of this class, and keep it
  - call `"new()"` on the class
  - wrenGetSlotHandle to get the new instance
- check if this class has a function like Update, if it does
  - add to Updateable list

Then each tick, you know everything in Updateable has `Update`, so you can call it. Same for Initable, and so on.

There's some details here to consider, the most important of all is that earlier note about reentrancy. If you get deeper into this method of implementing, don't try to call tick() from script. It has to originate in the c++ side.

**Comment by underscorediscovery** [comment]
Sure thing. Questions are never not welcome.
